### ISSUE TITLE
Compatibility with vtk 9.1 and higher

### ISSUE THREAD

**Issue by ealbiter** [ElmerCSC/elmerfem#532]
I tried compiling Elmer using VTK v9.3.1 in Arch Linux but ran into trouble.  After a quick search, I found that the file `vtkConfigure.h` had been [deprecated since v9.1](https://gitlab.kitware.com/vtk/vtk/-/blob/master/Documentation/release/9.1.md#filters-2).  I also found that this issue was spotted earlier in #308, but the modifications were not merged due to compatibility issues with older versions of VTK. 
Based on the PR, I made this patch that should be backward compatible:

```
diff --unified --recursive --text a/ElmerGUI/Application/cad/cadview.h b/ElmerGUI/Application/cad/cadview.h
--- a/ElmerGUI/Application/cad/cadview.h	2024-08-20 14:30:05.855951677 -0600
+++ b/ElmerGUI/Application/cad/cadview.h	2024-08-21 13:48:40.130823655 -0600
@@ -58,11 +58,14 @@
 #include <BRepMesh_IncrementalMesh.hxx>
 #endif
 
-#include "vtkConfigure.h"
 #ifndef VTK_MAJOR_VERSION
 #include "vtkVersionMacros.h"
 #endif
 
+#if VTK_MAJOR_VERSION <= 9 && VTK_MINOR_VERSION < 1
+#include "vtkConfigure.h"
+#endif
+
 class QMenu;
 class QAction;
 #if VTK_MAJOR_VERSION >= 8
diff --unified --recursive --text a/ElmerGUI/Application/vtkpost/vtkpost.h b/ElmerGUI/Application/vtkpost/vtkpost.h
--- a/ElmerGUI/Application/vtkpost/vtkpost.h	2024-08-20 14:30:05.875951676 -0600
+++ b/ElmerGUI/Application/vtkpost/vtkpost.h	2024-08-21 13:48:59.070823122 -0600
@@ -57,11 +57,14 @@
 #include <gui/PythonQtScriptingConsole.h>
 #endif
 
-#include "vtkConfigure.h"
 #ifndef VTK_MAJOR_VERSION
 #include "vtkVersionMacros.h"
 #endif
 
+#if VTK_MAJOR_VERSION <= 9 && VTK_MINOR_VERSION < 1
+#include "vtkConfigure.h"
+#endif
+
 class EpMesh;
 class ScalarField;
 #if VTK_MAJOR_VERSION >= 8
```
So far, so good. However, now I have a different error regarding `matc`:
```
/usr/bin/ld: listheaders: TLS definition in matc/src/libmatc.so section .tbss mismatches non-TLS reference in /tmp/ccsoZVGb.ltrans30.ltrans.o
/usr/bin/ld: /tmp/ccsoZVGb.ltrans30.ltrans.o: error adding symbols: bad value
collect2: error: ld returned 1 exit status
```
By the way, this error is not thrown when the compilation is made without VTK. I am currently trying to pinpoint the problem, but I haven't had any luck.

Any insights?

Cheers

**Comment by mmuetzel** [comment]
Thank you for your report and the proposed changes.

If I understand the release notes of VTK 9.1 that you linked correctly, this header was split into multiple finer-grained headers in that version.

Is it not necessary to include any of the finer-grained headers for newer versions of VTK? If it isn't, do you know why it was necessary to include that header with older versions of VTK?
Can including that header just be dropped for any version of VTK?

**Comment by ealbiter** [comment]
Thank you for your comment
> Is it not necessary to include any of the finer-grained headers for newer versions of VTK?

That is not necessary. I already tried the compilation using all the headers listed in the release notes, but I still got the same error. By the way, the error is only present when the compilation is done with both `MATC` and `VTK`. The error is not thrown if the compilation is performed with just one of them.

> If it isn't, do you know why it was necessary to include that header with older versions of VTK?
Can including that header just be dropped for any version of VTK?

Unfortunately, I have no idea, but I was wondering the same thing. Maybe @raback could help us with this.

**Comment by mmuetzel** [comment]
I opened #537 for the build issue with VTK 9.3 or newer. That should be addressing the first issue that you reported.
It doesn't touch the seconds issue (with MATC in ElmerGUI) yet.

**Comment by ealbiter** [comment]
I can confirm that commit 6eef153c77d0a9da9181a47b9369c06856abafec solves the problem with `matc`. Also, the change proposed in #537 fixes my compilation with `vtk` 9.3. It remains to see if that does not break something in older versions of VTK(?), as commented in #308.

              This cmake stuff is not really my cup of tea. I tested this on some platforms and happily merged only to find out that on an older one this would not compile. Hence I had to take this back. Trying to cut some corners didn't really pay of here. Sorry for the extra trouble!

_Originally posted by @raback in https://github.com/ElmerCSC/elmerfem/issues/308#issuecomment-985069317_
